### DISCUSSION TITLE
My response

### DISCUSSION THREAD

**Discussion by zardoy** [jimmywarting/you-might-not-need-typescript#3] [General]
Originally posted in https://github.com/jprichardson/node-jsonfile/pull/147#issuecomment-974880375

Hmm... this article is insane. However `.ts` files still have better syntax. Here is more comprehensive response:
> nobody likes wasting time compiling stuff.

Everybody knows how slow TSC is (I assume its the slowest compiler in the world). esbuild've had an article which described why so, tsc traverses through the ast of every file like for 5 times: to setup types infering -> linking etc.

However, nobody uses tsc at development, there is esbuild which can compile TypeScript project at any size in a few ms. And the typechecking of full project handling to the CI (or to lint-staged hook).

Just look at Type Guard section, which code looks more concise? I always spend time on setuping esbuid, just to have pretty and convenient syntax for inlining types everywhere. Its easy to do and you can use nice looking and structered code. I agree, that vscode has good suport for writing types in jsdocs though, but most extensions don't understand that.

Some other notes:

- `javascript.suggest.autoImports` isn't it already enabled by default?

- *Annotating destructured parameters* is wrong, it should be: `const { userName, age } = getUser() as DestructuredUser`. You can write it without moving your caret up-n-down and even without moving backward with TS postfixes.

> Notice how this requires you to write the same properties
> * twice to get the same function signature, so it's more verbose

Yes, it's annoying but it seems nobody knows you should instead:

1. Write properties as arguments
2. Use refactor action, which would save a lot of time to you:
![image](https://user-images.githubusercontent.com/46503702/142773682-6ece41eb-669b-4bb8-b9b8-d5b59abc630b.png)
Becomes
```ts
export function logUser({ userName, age, role, metadata }: { userName: string; age: number; role: 'mod' | 'admin'; metadata: SomethingElse }) {
    // ...
}
```
3. Then you should obviously extract it to interface for resuse or leave as is.

The same applies to jsdocs parameters, you shouldn't repeat yourself:

```ts
// <- hit ctrl+space here
const transformCode = (code: string, enableSomething: boolean) => {}
```
You got snippet:
```ts
/**
 * 
 * @param code 
 * @param enableSomething 
 */
const transformCode = (code: string, enableSomething: boolean) => {}
```
vscode will handle futher arg renames (but not adding new)

> Generics in JSDoc

I don't agree with you at this point. It is readable enough (if you add some spaces between `&` of course) since parameters are inlined. And as always with TypeScript you write it faster.

> You can even annotate on arrow functions

It would be worth to mention that arrows function shouldn't be used as `javascript.suggest.completeFunctionCalls` doesn't work with them.

> Things TypeScript gets wrong

Didn't get your point here. The same things applies to `//@ts-check`, otherwise you don't get the typechecking.

My friend also blamed on TypeScript, like he said "why people even use typescript?":

![image](https://user-images.githubusercontent.com/46503702/142775070-44b2cadf-4b33-47dd-85ab-0eed3f19c0fd.png)

Another example with generics:

![image](https://user-images.githubusercontent.com/46503702/142775176-fd9a0454-b75b-48b2-b1bc-99f793d50c2f.png)

And another example:

![image](https://user-images.githubusercontent.com/46503702/142775271-984d6361-bcd8-401b-acca-2c0fa010275e.png) Did I mean string?

(sorry for putting screenshots here)

Error on `1 + true`: its like to blame why I can't use promises in if statements (because you shouldn't!). TypeScript just want you to be explicit on `+` and other places where you expect implicit type converstion. I've heard about flag request that disables checks on binary operators.

Error on alert is also make sense: just don't use alert. We've got amazing `console` thing in JS.

`signal` I fixed by using custom lib.d.ts.

However `replaceAll` is a special one.

One of the stupidest thing of TypeScript is stupid `tsconfig.json` defaults. I recommend to never use `tsconfig` (or jsconfig) without shared configuration. Examples:
https://github.com/sindresorhus/tsconfig
My own is a beast: https://github.com/zardoy/tsconfig

And create some kind of command so you can quickly add it easily to every just created project e.g. [something like with mrm](https://github.com/zardoy/mrm-preset/blob/main/typescript/index.ts).

There are a lot of things, that TypeScript docs will never share with you (such as [make tech](https://github.com/zardoy/vscode-framework/blob/5f704727f6165a6946b73cf290f446d4d8a1548c/packages/vscode-framework/src/cli/manifest-generator/manifestGenerators.ts#L43-L47): another time-saving thing). There are a lot of things that will most probably never be fixed such as inability of TypeScript compiler to remove previous output files. Projects with `composite: true` would save a lot of time in monorepos by using incremetal build, but it is also requires [workaround](https://github.com/zardoy/vscode-framework/blob/5f704727f6165a6946b73cf290f446d4d8a1548c/scripts/prepare.mjs). I hate tsc and using it only in for building libraries, but even at this point it can do bad things: https://github.com/microsoft/TypeScript/issues/42853

But here what about building libraries:

Libraries that utilizes types in jsdocs can have really good typings support, such as `unist` ecosystem.
Another example is `phaser` that generated typings from jsdocs, but then migrated to pure `.d.ts` (as I remember because it was hard to link types).

And finally what about this library. I published `typed-jsonfile` (it doesn't have readme, but you can look at api at https://paka.dev/npm/typed-jsonfile), that have helpers (for writing `package.json` files for example):

![image](https://user-images.githubusercontent.com/46503702/142775678-86fd4959-fd31-424e-91e2-6ab73c366c66.png)

Isn't typescript ecosystem is cool? I hope you now understand what was the intention of my comment above.

Ok, I'm done at this point.

Though, thank you very much for sharing the article! The quality is extremely good. Thank you for using monaco editors.

**Comment by jimmywarting** [comment]
Ty for being engaged and reviewing my article. And i'm glad that you thought it had good quality. 

I don't think typescript is all that bad. I mean... i use it myself in a sense. So i tough I would throw that mush trash at it. Tried to make it like one of that "you might not need javascript/jquery/momentjs" artikels. heck i even use [typedoc](https://github.com/TypeStrong/typedoc) for a pure js project with some jsdoc cuz typescript have better intelligence and can figure out the return value with using `@return` or `function x(): string` for typescript.
It's far better than any jsdoc-to-html generator.

I'm thankful for everyone who writes with types support, cuz it helps a lot. 

I have known javascript for 12 years before typescript came a long and since then I have seen other compile-to-js language come and go, that is why haven't been sold on the typescript syntax, and i have tried it for about one year professionally at work and have never really fallen in love for its wired syntax.

**Comment by jimmywarting** [comment]
> Everybody knows how slow TSC is (I assume its the slowest compiler in the world). esbuild've had an article which described why so, tsc traverses through the ast of every file like for 5 times: to setup types infering -> linking etc.
> 
> However, nobody uses tsc at development, there is esbuild which can compile TypeScript project at any size in a few ms. And the typechecking of full project handling to the CI (or to lint-staged hook).

It is not just about compiling tsc when building for production, you also waste lot's of initial startup time when using [ts-node](https://github.com/TypeStrong/ts-node), [ts-node-dev](https://www.npmjs.com/package/ts-node-dev) or similar, i'm well aware that tsc is mostly just the end station before you hit publish. 

The way you hook into node's runtime to add support for running/executing typescript inside of node is all but an extra runtime compiler that needs to parse the javascript twice from ts > js > machine code...

deno stop using it after all... because it took time to compile every time you want do some development https://startfunction.com/deno-will-stop-using-typescript/

**Reply by jimmywarting** [reply]
now when you want to support esm and has to hook into nodes experimental module loader then this becomes even complicated

**Reply by zardoy** [reply]
Just forget about ts-node and use [tsm](https://www.npmjs.com/package/tsm)

**Reply by zardoy** [reply]
It is fast just like node. `node file.js` -> `tsm file.ts`

**Reply by jimmywarting** [reply]
haven't tried or heard about that, but thx for the suggestion, looks promising for a work related thing where we use TS + esm

**Reply by zardoy** [reply]
I'm using it for like 2 months and it's a beast. It also provides sourcemaps, so you can obviously run `tsm file.ts` in JavaScript Debug Terminal in VSCode and breakpoints will work properly.

**Reply by jimmywarting** [reply]
when we adopted to esm then ts-node-dev wouldn't play ball. 
going to suggest tsm to the team and have a look at it

**Reply by zardoy** [reply]
What about Deno.
Deno as I remember watned to rewrite some parts of TypeScript Compiler on Rust https://github.com/denoland/deno/issues/5432. And from this threat I found [estrella](https://github.com/rsms/estrella), which should make developing large TypeScript projects a little easier (in the opposite to `tsm` it still performs typechecking).

**Reply by jimmywarting** [reply]
I don't get it what all the hype is about typescript, Why not just code stuff in rust, c or any other language and compile it to wasm if one are that into types, go and get that native speed...!?

Typescript is just a layer on top of javascript and typescript is never going to be part of the browser or node's official runtime and it will like i said: always stay one step behind in new features", so you always going to have to deal with compiling stuff to javascript no matter what if you want stuff to run in deno, node and browsers... [Google tried and fail to impl dart](https://stackoverflow.com/a/8947021/1008999)
simple reason being "don't want to have a bloated multi vm/runtime compiler"
I think this SO post should be a solo reason of why you should bet on javascript (or wasm), maybe in the feature another compile-to-js will be hype or heck: we might even get [optional static typing](https://github.com/sirisian/ecmascript-types) 

what would your/deno opinion be of typescript then - if we got optional static typing?

**Comment by jimmywarting** [comment]
> Generics in JSDoc
> 
> I don't agree with you at this point. It is readable enough (if you add some spaces between & of course) since parameters are inlined. And as always with TypeScript you write it faster.

I think it's still a mouth full too look at by just glaring at:
`const deleteName = <T>(obj: T & { name?: string; }): Omit<T,'name'> => {`

I think it's only weird symbols everywhere, i have seen worse... and this is small example. Honestly, i don't think everything need to be typed just for the sake of having types, i guess i wouldn't have written the jsdoc at all unless it was going to bring me any valuable support, but here I just added it to show you that it's possible with jsdoc as well.

It would depend if it was internal or public... reading `const deleteName = (obj) => {` is so much more cleaner code. the jsdoc solution isn't the fanciest either but at least it document each row/thing separately. i can agree that both are not so perfect

**Comment by jimmywarting** [comment]
> Things TypeScript gets wrong
> Didn't get your point here. The same things applies to //@ts-check, otherwise you don't get the typechecking.

I know :P 
But there is so many things i think that typescript should not whine about... 
and it don't always support the newest features. which brings me to the next comment:

> signal I fixed by using custom lib.d.ts.

I don't think you should have to do it. i wished it could be less strict... TypeScript dose not know everything. it would be nice to be able to do things typescript dose not understand without giving errors here and there. and adding `@ts-ignore` to things

**Reply by jimmywarting** [reply]
One resent thing i wanted to do today was `new TextEncoder().encode(new URLSearchParams({...}))`

once again, TypeScript don't understand that the default behaviour is to cast whatever to string so i had to use `.toString()`

**Comment by jimmywarting** [comment]
> Error on alert is also make sense: just don't use alert. We've got amazing console thing in JS.

The point where not about using alert vs console
It was about using extra arguments :P

**Comment by jimmywarting** [comment]
> but most extensions don't understand that.

For instances? I would like to be challenged that things can be done in a vanilla js way :P  

> javascript.suggest.autoImports isn't it already enabled by default?

I don't know...

> My own is a beast

I got one of my [own](https://github.com/jimmywarting/native-file-system-adapter/blob/master/tsconfig.json)  that i use in my projects
...even do i don't use typescript syntax, guess i should change it to jsconfig.json :P 
I don't use it as a "extend" 
it's in my private skeleton-template-repo that i start of from
